### PR TITLE
Weblogic empty servlet path fallback

### PR DESCRIPTION
**Pull request by SylvainJuge**
## What does this PR do?

Improves Servlet path naming on Weblogic when `use_path_as_transaction_name=true` is set.

## Checklist

- [x] This is a bugfix
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [x] I have added tests that would fail without this fix

### CODE REVIEW

File: apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java [hunk 1]
```diff
@@ -239,10 +239,26 @@ public static <HttpServletRequest, HttpServletResponse, ServletContext, ServletC
...
                 }
 
-                servletTransactionHelper.onAfter(transaction, t == null ? t2 : t, adapter.isCommitted(httpServletResponse), adapter.getStatus(httpServletResponse),
-                    overrideStatusCodeOnThrowable, adapter.getMethod(httpServletRequest), parameterMap, adapter.getServletPath(httpServletRequest),
-                    adapter.getPathInfo(httpServletRequest), contentTypeHeader, true
-                );
+                ServletContext servletContext = adapter.getServletContext(httpServletRequest);
+                String servletPath = adapter.getServletPath(httpServletRequest);
+                String pathInfo = adapter.getPathInfo(httpServletRequest);
+                if (servletPath == null || servletPath.length() == 0 && servletContext != null) {
```
Comment (felixbarny) [inline apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java:245]:
```suggestion
                if ((servletPath == null || servletPath.length() == 0) && servletContext != null) {
```

IINM, it would be equivalent to `if (servletPath == null || (servletPath.length() == 0 && servletContext != null))` otherwise.

File: apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java [hunk 2]
```diff
@@ -239,10 +239,26 @@ public static <HttpServletRequest, HttpServletResponse, ServletContext, ServletC
...
+                    String contextPath = adapter.getContextPath(servletContext);
+                    String requestURI = adapter.getRequestURI(httpServletRequest);
+                    servletPath = servletTransactionHelper.normalizeServletPath(requestURI, contextPath, servletPath, pathInfo);
+                }
+
+                servletTransactionHelper.onAfter(
```
Comment (felixbarny) [inline apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java:251]: Much cleaner to have every argument in a separate line. I like!

File: apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletTransactionHelper.java [hunk 3]
```diff
@@ -144,13 +144,38 @@ public static void setUsernameIfUnset(@Nullable String userName, TransactionCont
         }
     }
 
+    public String normalizeServletPath(String requestURI, @Nullable String contextPath, @Nullable String servletPath, @Nullable String pathInfo) {
+        String path = servletPath;
+        if (path == null || path.isEmpty()) {
```
Comment (felixbarny) [inline apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletTransactionHelper.java:149]: We'd expect this case to be relatively rare, right? It does cause some allocations but if it's an edge case I think that's fine.

File: apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java (hunk 1 above)
Comment (SylvainJuge) [reply apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java:245]: Good catch!

File: apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletTransactionHelper.java (hunk 3 above)
Comment (SylvainJuge) [reply apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletTransactionHelper.java:149]: Yes, this is definitely an edge case as most of the time we have `servletPath` provided as we usually expect.

File: integration-tests/simple-webapp/src/main/webapp/WEB-INF/web.xml [hunk 4]
```diff
@@ -27,6 +27,15 @@
         <url-pattern>/servlet</url-pattern>
     </servlet-mapping>
 
+    <servlet-mapping>
+        <!--
+        allows testing servlet mapping with wildcard, can make servlet path to be empty/null according to spec
+        but in practice this behavior relies on app server implementation and might differ.
+        -->
+        <servlet-name>TestServlet</servlet-name>
+        <url-pattern>/test-servlet/*</url-pattern>
```
Comment (felixbarny) [inline integration-tests/simple-webapp/src/main/webapp/WEB-INF/web.xml:36]: Does this reproduce the exact issue we're seeing on Weblogic?

Comment (SylvainJuge) [reply integration-tests/simple-webapp/src/main/webapp/WEB-INF/web.xml:36]: Not directly, if I understand the servlet spec correctly this should in theory make `servletPath` to be empty, but in practice with Tomcat this is not the case. for example we get `servletPath` = `/test-servlet` and `pathInfo` = `/info` when request URL is `/test-servlet/info` which is also a bit confusing. I am not sure it's really worth adding an integration test for that case though, I've kept it mostly for manual testing.

**Review by felixbarny — APPROVED** [review]
LGTM, but let's wait for the feedback before merging.

File: apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletTransactionHelper.java [hunk 5]
```diff
@@ -146,30 +146,35 @@ public static void setUsernameIfUnset(@Nullable String userName, TransactionCont
...
+        if (path != null && !path.isEmpty()) {
+            return path;
+        }
 
-            if (start == end) {
-                // use complete request URI instead of an empty string as fallback
-                path = requestURI;
-            } else {
-                path = requestURI.substring(start, end);
+        logger.debug("Empty servlet path fallback applied. requestURI = {}, contextPath = {}, servletPath = {}, pathInfo = {}", requestURI, contextPath, servletPath, pathInfo);
```
Comment (felixbarny) [inline apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletTransactionHelper.java:153]: nit: I'd guard this with a if so that the varargs array is only allocated when the debug log level is enabled.
